iree/compiler/Conversion/LinalgToLLVM/ConvertToLLVM.cpp

```diff
@@ -0,0 +1,159 @@
+ public:
+  explicit ConvertRankedDimPattern(MLIRContext *context,
+                                   LLVMTypeConverter &lowering_)
+      : ConvertToLLVMPattern(Shape::RankedDimOp::getOperationName(), context,
+                             lowering_) {}
+
+  LogicalResult matchAndRewrite(
+      Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override {
+    auto rankedDimOp = cast<Shape::RankedDimOp>(op);
+    auto makeRankedShapeOp =
+        cast<Shape::MakeRankedShapeOp>(rankedDimOp.shape().getDefiningOp());
+    if (!makeRankedShapeOp) return failure();
+    auto dimIndex = rankedDimOp.index();
+    auto dynamicDims =
+        makeRankedShapeOp.dynamic_dimensions()[*dimIndex.getRawData()];
+    rewriter.replaceOp(op, dynamicDims);
+    return success();
+  }
+};
+
+class RemoveMakeRankedShape : public ConvertToLLVMPattern {
+ public:
+  explicit RemoveMakeRankedShape(MLIRContext *context,
+                                 LLVMTypeConverter &lowering_)
+      : ConvertToLLVMPattern(Shape::MakeRankedShapeOp::getOperationName(),
+                             context, lowering_) {}
+
+  LogicalResult matchAndRewrite(
+      Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override {
+    rewriter.eraseOp(op);
+    return success();
+  }
+};
+
+namespace {
+struct ConvertToLLVMPass
+    : public PassWrapper<ConvertToLLVMPass, OperationPass<ModuleOp>> {
+  void runOnOperation() override;
+};
+
+}  // namespace
+
+void ConvertToLLVMPass::runOnOperation() {
+  auto module = getOperation();
+  OwningRewritePatternList patterns;
+  LLVMTypeConverter converter(&getContext());
+  populateAffineToStdConversionPatterns(patterns, &getContext());
+  populateLoopToStdConversionPatterns(patterns, &getContext());
+  populateStdToLLVMConversionPatterns(converter, patterns);
+  populateVectorToSCFConversionPatterns(patterns, &getContext());
+  populateVectorToLLVMMatrixConversionPatterns(converter, patterns);
+  populateVectorToLLVMConversionPatterns(converter, patterns);
+  patterns.insert<ConvertRankedDimPattern, ConvertTieShapePattern,
+                  RemoveMakeRankedShape>(&getContext(), converter);
+  populateLinalgToLLVMConversionPatterns(converter, patterns, &getContext());
+
+  LLVMConversionTarget target(getContext());
+  target.addDynamicallyLegalOp<FuncOp>(
+      [&](FuncOp op) { return converter.isSignatureLegal(op.getType()); });
+  target.addLegalOp<ModuleOp, ModuleTerminatorOp>();
+  if (failed(applyPartialConversion(module, target, patterns, &converter)))
+    signalPassFailure();
+}
+
+std::unique_ptr<OperationPass<ModuleOp>> createConvertToLLVMPass() {
+  return std::make_unique<ConvertToLLVMPass>();
+}
+
+static PassRegistration<ConvertToLLVMPass> pass(
+    "iree-codegen-convert-to-llvm",
+    "Perform final conversion from Linalg/HAL/Shape/Vector/Standard to LLVMIR "
+    "dialect",
+    [] { return std::make_unique<ConvertToLLVMPass>(); });
+
```

Comment:
Can we expose this via pass registration and have a test mlir for this pass? This pass is quite important to get correctly (ABI mismatch is terrible to debug) so worth some extra test coverage. Otherwise we only see integration tests failing. Tests themselves can also be good documentation of the logic of the pass. :)